rfcs/0007-handling-corruption.md

```diff
@@ -0,0 +1,392 @@
+
+Since `get` calls may be ongoing during corruption, it is possible that the
+callback will be called continuously and by multiple threads. Therefore, any
+user-defined callback must be thread-safe. Additionally, the callback should
+avoid slow or blocking operations to ensure the database remains available if
+corruption is encountered.
+
+To support operators who value durability, the developer making the `get` call
+can choose to abort the application if a corruption error is returned. The
+application can then provide operators with detailed information included in
+the error returned by SlateDB, including the nature of the corruption and
+suggestions on how to fix it.
+
+##### Remaining Available
+In the case where corruption is encountered, and `DBOptions.on_corruption` allows
+operation to continue, SlateDB should endeavor to return the most recent 
+valid data it can.
+
+- If the SST Bloom Filter indicates a value exists in the SST, yet the key was
+  not found, and corruption was detected during iteration, the `get` caller
+  should receive a Corruption error.
+- If the most recent value is in a non-corrupt block, the `get` call should
+  return successfully
+- If the SST index is corrupt and `get` has not yet found the KV in an earlier
+  SST, the call should receive a Corruption error. (This effectively prevents
+  reads in SSTs older than the first corrupt SST)
+- MORE?
+
+#### Corruption Inspection and Repair
+To support the repair of SSTs, I propose introducing two new functions:
+`sst_inspect()` and `sst_repair()`. These functions will provide developers with
+tools to inspect the current contents of a specific SST, and repair a specific
+SST that has been identified as corrupt.
+
+To support the repair of the WAL, I propose introducing two additional functions:
+`wal_inspect()` and `wal_repair()`. These functions will allow developers to
+inspect the current contents of the WAL, including corrupted entries, and to
+repair the WAL to the best of its ability through entity recovery or truncation
+of the WAL.
+
+The inspection functions are designed to provide verbose information about the
+WAL or SST. This includes reporting the contents of the files and any
+corruption found. The goal is to provide operators with tools to inspect the 
+file contents, any corruption, and the surrounding data without needing intimate
+knowledge of the file format.
+
+The inspect and repair functions should share the following characteristics:
+- Both WAL and SST variants should accept raw file input, such that procurement
+  of the file via SST ID (either WAL or compacted) or out of band file procurement
+  is possible. This allows operators or developers to download files out of band
+  and inspect or repair the files using the provided functions
+- Repair functions should include a "dry run" option which only reports repairs 
+without actually perform the repair.
+
+We should introduce a new function called `repair()`, which encapsulates both
```

Comment:
An admin command that checks consistency of the DB, by checking all the SSTs for the list of corruptions that can be identified could be useful, it could be run at some intervals.